evm/p2p/lightchain.py

```diff
@@ -0,0 +1,76 @@
+import asyncio
+import logging
+
+from async_lru import alru_cache
+
+from eth_utils import encode_hex
+
+from evm.chains import Chain
+from evm.exceptions import (
+    BlockNotFound,
+)
+from evm.p2p.utils import gen_request_id
+
+
+class OnDemandDataBackend:
+    logger = logging.getLogger("evm.p2p.lightchain.OnDemandDataBackend")
+    # TODO:
+    # 1. Use the PeerManager to maintain a set of connected peers
+    # 2. Implement a queue of requests and a distributor which picks items from that queue and
+    # sends them to one of our peers, ensuring the selected peer has the info we want and
+    # respecting the flow control rules
+
+    @alru_cache(maxsize=1024)
+    @asyncio.coroutine
+    def get_block_by_hash(self, block_hash):
+        peer = yield from self.get_peer()
+        self.logger.debug("Fetching block {} from peer {}".format(encode_hex(block_hash), peer))
+        request_id = gen_request_id()
+        peer.les_proto.send_get_block_bodies([block_hash], request_id)
+        reply = yield from peer.wait_for_reply(request_id)
+        if len(reply['bodies']) == 0:
+            raise BlockNotFound("No block with hash {} found".format(block_hash))
+        return reply['bodies'][0]
+
+    @asyncio.coroutine
+    def get_peer(self):
+        raise NotImplementedError("TODO")
+
+    @asyncio.coroutine
+    def stop(self):
+        raise NotImplementedError("TODO")
+
+
+class LightChain(Chain):
+    on_demand_data_backend_class = None
+
+    def __init__(self, chaindb, header=None):
+        super(LightChain, self).__init__(chaindb, header=header)
+        self.on_demand_data_backend = self.on_demand_data_backend_class(chaindb)
+
+    @asyncio.coroutine
+    def stop(self):
+        yield from self.on_demand_data_backend.stop()
+
+    @asyncio.coroutine
+    def get_canonical_block_by_number(self, block_number):
+        try:
+            block_hash = self.chaindb.lookup_block_hash(block_number)
+        except KeyError:
+            raise BlockNotFound("No block with number {} found on canonical chain".format(
```

Comment:
Making sure I understand what is going on here:

This uses `self.chaindb` rather than `self.on_demand_data_backend` because it needs to return the *local* version of what the canonical block number is.  *If* we are still syncing block headers and we request a block number that is higher than the latest synced one, we'll hit this exception.

Does it make sense to have a secondary code path here for when `block_number` is less than what we *believe* the current highest block number is, but still greater than the latest block we've synced?

Reply:
For simplicity the LightChain currently assumes the headers and canonical block numbers are always synced, which is probably ok for now as it's only the initial sync that may take some time (15min or so), and after that the headers are synced almost instantaneously every time we get a new announce msg. If we could (and we should be able to) speed up the initial sync, I think it'd be fine to keep the LightChain always assuming headers are fully synced, for the time being, at least